File: dotnet/src/CdaToFhirAdapter.Core/Parsers/CdaParser.cs
using System.Xml;
using System.Xml.Linq;
using CdaToFhirAdapter.Core.Exceptions;

namespace CdaToFhirAdapter.Core.Parsers;

/// <summary>
/// Parses HL7 CDA R2 XML documents into an <see cref="XDocument"/> and exposes
/// XPath-style helper methods for field extraction.
/// </summary>
/// <remarks>
/// The parser is hardened against XXE attacks by using
/// <see cref="XmlResolver"/>-disabled settings.
/// </remarks>
public sealed class CdaParser
{
    private static readonly XNamespace Hl7Ns = "urn:hl7-org:v3";

    /// <summary>
    /// Parses the supplied CDA XML string into an <see cref="XDocument"/>.
    /// </summary>
    /// <param name="cdaXml">CDA document as an XML string.</param>
    /// <returns>The parsed <see cref="XDocument"/>.</returns>
    /// <exception cref="CdaParseException">Thrown when the XML is null, blank, or malformed.</exception>
    public XDocument Parse(string cdaXml)
    {
        if (string.IsNullOrWhiteSpace(cdaXml))
            throw new CdaParseException("CDA XML must not be null or blank.");

        try
        {
            var settings = new XmlReaderSettings
            {
                DtdProcessing = DtdProcessing.Prohibit,
                XmlResolver = null,
                MaxCharactersFromEntities = 0
            };

            using var reader = XmlReader.Create(new StringReader(cdaXml), settings);
            return XDocument.Load(reader);
        }
        catch (XmlException ex)
        {
            throw new CdaParseException($"Failed to parse CDA XML: {ex.Message}", ex);
        }
    }

    /// <summary>Returns the first matching element value, or an empty string.</summary>
    public string GetElementValue(XDocument doc, params string[] localNames)
    {
        XElement? current = doc.Root;
        foreach (var name in localNames)
        {
            current = current?.Element(Hl7Ns + name);
            if (current is null) return string.Empty;
        }
        return current?.Value.Trim() ?? string.Empty;
    }

    /// <summary>Returns the first matching element, or null.</summary>
    public XElement? GetElement(XDocument doc, params string[] localNames)
    {
        XElement? current = doc.Root;
        foreach (var name in localNames)
        {
            current = current?.Element(Hl7Ns + name);
            if (current is null) return null;
        }
        return current;
    }

    /// <summary>Returns a named attribute value from the first matching element, or empty string.</summary>
    public string GetAttribute(XDocument doc, string attributeName, params string[] localNames)
    {
        var element = GetElement(doc, localNames);
        return element?.Attribute(attributeName)?.Value.Trim() ?? string.Empty;
    }

    /// <summary>Returns all direct children with a given local name under a path.</summary>
    public IEnumerable<XElement> GetElements(XDocument doc, params string[] localNames)
    {
        XElement? current = doc.Root;
        for (int i = 0; i < localNames.Length - 1; i++)
        {
            current = current?.Element(Hl7Ns + localNames[i]);
            if (current is null) return Enumerable.Empty<XElement>();
        }
        return current?.Elements(Hl7Ns + localNames[^1]) ?? Enumerable.Empty<XElement>();
    }
}

File: java/src/main/java/ie/hl7/cda/fhir/parser/CdaParser.java
package ie.hl7.cda.fhir.parser;

import ie.hl7.cda.fhir.exception.CdaParseException;
import org.springframework.stereotype.Component;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import javax.xml.XMLConstants;
import javax.xml.namespace.NamespaceContext;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;
import java.io.IOException;
import java.io.StringReader;
import java.util.Iterator;

/**
 * Parses HL7 CDA R2 XML documents into a DOM {@link Document} and exposes
 * XPath-based accessors for field extraction.
 *
 * <p>The parser is secured against XXE attacks by disabling external entity
 * and DTD processing.</p>
 */
@Component
public class CdaParser {

    /** HL7v3 / CDA namespace URI. */
    public static final String HL7_NS = "urn:hl7-org:v3";

    /** Namespace prefix used in XPath expressions. */
    private static final String NS_PREFIX = "cda";

    /**
     * Parses the supplied CDA XML string into a DOM Document.
     *
     * @param cdaXml the CDA document as an XML string
     * @return the parsed DOM Document
     * @throws CdaParseException if the XML is malformed or entity expansion is detected
     */
    public Document parse(final String cdaXml) {
        if (cdaXml == null || cdaXml.isBlank()) {
            throw new CdaParseException("CDA XML must not be null or blank");
        }
        try {
            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            factory.setNamespaceAware(true);
            // Disable XXE
            factory.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
            factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
            factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
            factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
            factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
            DocumentBuilder builder = factory.newDocumentBuilder();
            return builder.parse(new InputSource(new StringReader(cdaXml)));
        } catch (ParserConfigurationException | SAXException | IOException ex) {
            throw new CdaParseException("Failed to parse CDA XML: " + ex.getMessage(), ex);
        }
    }

    /**
     * Evaluates an XPath expression against the given document and returns the
     * text content of the first matching node.
     *
     * @param doc    source DOM document
     * @param xpathExpression XPath expression using {@code cda:} prefix for the HL7 namespace
     * @return trimmed text content, or an empty string if no match
     */
    public String extractText(final Document doc, final String xpathExpression) {
        try {
            XPath xpath = buildXPath();
            Node node = (Node) xpath.evaluate(xpathExpression, doc, XPathConstants.NODE);
            return node == null ? "" : node.getTextContent().trim();
        } catch (XPathExpressionException ex) {
            return "";
        }
    }

    /**
     * Evaluates an XPath expression and returns the attribute value of the
     * first matching node.
     *
     * @param doc             source DOM document
     * @param xpathExpression XPath expression
     * @param attribute       attribute name to read
     * @return attribute value, or an empty string if no match
     */
    public String extractAttribute(final Document doc, final String xpathExpression,
            final String attribute) {
        try {
            XPath xpath = buildXPath();
            Node node = (Node) xpath.evaluate(xpathExpression, doc, XPathConstants.NODE);
            if (node == null) {
                return "";
            }
            Node attr = node.getAttributes().getNamedItem(attribute);
            return attr == null ? "" : attr.getNodeValue().trim();
        } catch (XPathExpressionException ex) {
            return "";
        }
    }

    /**
     * Returns a NodeList for the given XPath expression.
     *
     * @param doc             source DOM document
     * @param xpathExpression XPath expression
     * @return matching NodeList (may be empty)
     */
    public NodeList extractNodes(final Document doc, final String xpathExpression) {
        try {
            XPath xpath = buildXPath();
            return (NodeList) xpath.evaluate(xpathExpression, doc, XPathConstants.NODESET);
        } catch (XPathExpressionException ex) {
            return new EmptyNodeList();
        }
    }

    private XPath buildXPath() {
        XPath xpath = XPathFactory.newInstance().newXPath();
        xpath.setNamespaceContext(new CdaNamespaceContext());
        return xpath;
    }

    /**
     * Namespace context that maps the {@code cda} prefix to the HL7v3 namespace.
     */
    private static final class CdaNamespaceContext implements NamespaceContext {

        @Override
        public String getNamespaceURI(final String prefix) {
            if (NS_PREFIX.equals(prefix)) {
                return HL7_NS;
            }
            return XMLConstants.NULL_NS_URI;
        }

        @Override
        public String getPrefix(final String namespaceURI) {
            if (HL7_NS.equals(namespaceURI)) {
                return NS_PREFIX;
            }
            return null;
        }

        @Override
        public Iterator<String> getPrefixes(final String namespaceURI) {
            return null;
        }
    }

    /**
     * Empty NodeList implementation used as a safe null-object return value.
     */
    private static final class EmptyNodeList implements NodeList {

        @Override
        public Node item(final int index) {
            return null;
        }

        @Override
        public int getLength() {
            return 0;
        }
    }
}
